pd/migrations/20211216025603_init.sql

```diff
@@ -113,30 +106,59 @@ CREATE TABLE IF NOT EXISTS delegation_changes (
 CREATE INDEX ON delegation_changes (epoch);
 CREATE INDEX ON delegation_changes (validator_identity_key);
 
--- Set of quarantined notes: once the epoch is large enough, all old-enough quarantined notes should
--- be inserted into the NCT and the notes table (and removed from this table)
+-- Nullifiers, indexed by height
+CREATE TABLE IF NOT EXISTS nullifiers (
+    nullifier bytea PRIMARY KEY,
+    height bigint NOT NULL REFERENCES blocks (height),
+
+    -- if this nullifier was quarantined at any point in the past:
+    unbonding_height bigint, -- height at which to make the spend permanent
+    reverted_height bigint REFERENCES blocks (height), -- height at which the revert was made, if it was reverted
+    validator_identity_key bytea REFERENCES validators (identity_key),
+
+    -- iff the nullifier was quarantined, it has an unbonding height and a validator identity key
+    CONSTRAINT valid_quarantine CHECK (
+        (unbonding_height IS NULL OR validator_identity_key IS NOT NULL) AND
+        (validator_identity_key IS NULL OR unbonding_height IS NOT NULL)
+    ),
+    -- if the nullifier has a reverted height, then it was quarantined, so it has an unbonding
+    -- height and a validator identity key
+    CONSTRAINT valid_reverted CHECK (
+        reverted_height IS NULL
+        OR (unbonding_height IS NOT NULL AND validator_identity_key IS NOT NULL)
+    ),
+    -- unbonding_height can't be negative
+    CONSTRAINT positive_unbonding_height
+        CHECK (unbonding_height IS NULL OR unbonding_height >= 0),
+    -- if unbonding_height is not null, it is greater than or equal to height
+    CONSTRAINT unbonding_height_greater_than_height
+        CHECK (unbonding_height IS NULL OR height <= unbonding_height),
+    -- if reverted_height is not null, it is less than or equal to unbonding_height
+    CONSTRAINT reverted_height_less_than_or_equal_unbonding_height
+        CHECK (reverted_height IS NULL OR reverted_height <= unbonding_height)
```

Comment:
These constraints are fine as a backup, but we should not be pushing our application logic into the database, and I worry that adding them now commits to maintaining them in the future.